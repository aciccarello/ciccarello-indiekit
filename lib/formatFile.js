// @ts-check
import { stringify } from "yaml";

const mastodonTitleUserRegex = /^(.*)\((@[\w\.]+@[\w\.]+)\)$/;
const mastodonUserProfileRegex = /^https?:\/\/[\w\.]+\/@[\w\.]+/;

export default async function formatFile(properties) {
  try {
    console.log(
      "Processing postTemplate",
      JSON.stringify(properties, undefined, "\t")
    );
    let postContent;
    if (properties.content) {
      postContent =
        properties.content.text ||
        properties.content.html ||
        properties.content;
      postContent = `${postContent}\n`;
    } else {
      postContent = "";
    }

    const {
      published,
      deleted,
      photo: photoProperty,
      references: referencesMap = {},
    } = properties;

    let photo = Array.isArray(photoProperty)
      ? photoProperty.shift()
      : photoProperty;

    const { url: image, alt: image_alt, ...image_other } = photo || {};

    // Try to normalize references to fix common errors on people's sites
    const references = Object.entries(referencesMap).map(([url, ref]) => {
      let entry = ref;
      if (!entry.type && entry.children) {
        // Find main entry
        entry = entry.children.find(({ type }) => type === "entry");
      }
      if (!entry) {
        console.warn("Entry not found on referenced page", ref.url, ref);
        return ref;
      }
      let { name, author, content, summary, "post-type": postType } = entry;
      if (!author) {
        // Assume top level card is the author
        author = entry.children?.find(({ type }) => type === "card");
        console.warn(author, entry);
      }
      if (!author) {
        // Assume top level card is the author
        author = ref?.children?.find(({ type }) => type === "card");
      }
      if (!author && name) {
        const authorNameParts = mastodonTitleUserRegex.exec(name);
        const authorProfileUrl = mastodonUserProfileRegex.exec(url);

        if (authorNameParts) {
          const [_, authorName, username] = authorNameParts;
          const [authorUrl] = authorProfileUrl ?? [undefined];

          author = {
            type: "card",
            name: authorName.trim(),
            nickname: username,
            url: authorUrl,
          };
          name = undefined;
        }
      }

      if (!content && summary) {
        content = summary;
      }
      if (content?.text) {
        // Don't need both the HTML and text representation of content
        content = content.text;
      }
      if (content?.length > 500) {
        content = content.slice(0, 500) + "â€¦";
      }

      // Whitelist what properties to include
      // Some sites include extra properties like comments that I don't need
      const { type, published, updated } = entry;

      return {
        url: ref.url ?? url,
        name,
        "post-type": postType,
        type,
        published,
        updated,
        author,
        // TODO: Should this be summary rather than content?
        content,
      };
    });

    properties = {
      date: published,
      updated: properties.updated,
      // TODO: Check how visibility is set
      eleventyExcludeFromCollections: deleted
        ? "deleted"
        : properties["post-status"] === "draft"
        ? "draft"
        : properties.visibility === "public"
        ? undefined
        : properties.visibility,
      // Ignore title from micropub-like endpoint
      title: ["like"].includes(properties["post-type"])
        ? undefined
        : properties.name,
      slug: properties["mp-slug"],
      "bookmark-of": properties["bookmark-of"],
      "like-of": properties["like-of"],
      "repost-of": properties["repost-of"],
      "in-reply-to": properties["in-reply-to"],
      tags: properties.category,
      image,
      image_alt,
      references,

      // Untested properties
      image_other: Object.keys(image_other).length ? image_other : undefined,
      indiekit_excerpt: properties.summary,
      indiekit_start: properties.start,
      indiekit_end: properties.end,
      indiekit_rsvp: properties.rsvp,
      indiekit_location: properties.location,
      indiekit_checkin: properties.checkin,
      indiekit_audio: properties.audio,
      indiekit_video: properties.video,

      "indiekit_post-type": properties["post-type"],
      syndication: properties.syndication,
      "mp-syndicate-to": properties["mp-syndicate-to"],
    };

    return `---
${stringify(properties)}
---

${postContent}
`;
  } catch (error) {
    console.error("Problem generating post file content", error, properties);
    throw new Error(
      "Unable to create post file contents. See logs for details"
    );
  }
}
