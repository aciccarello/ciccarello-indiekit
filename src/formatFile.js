// @ts-check
const { stringify } = require("yaml");

module.exports = function formatFile(properties) {
  try {
    console.log(
      "Processing postTemplate",
      JSON.stringify(properties, undefined, "\t")
    );
    let postContent;
    if (properties.content) {
      postContent =
        properties.content.text ||
        properties.content.html ||
        properties.content;
      postContent = `${postContent}\n`;
    } else {
      postContent = "";
    }

    const { published, deleted, photo, references } = properties;

    const { url: image, alt: image_alt, ...image_other } = photo || {};

    // Try to normalize references to fix common errors on people's sites
    for (const url in references) {
      let entry = references[url];
      if (!entry.type && entry.children) {
        // Find main entry
        entry = entry.children.find(({ type }) => type === "entry");
      }
      if (!entry) {
        console.warn("Entry not found on referenced page", url, entry);
        break;
      }
      let { author, content } = entry;
      if (!author) {
        // Assume top level card is the author
        author = references[url].children.find(({ type }) => type === "card");
      }
      if (content?.text) {
        // Don't need both the HTML and text representation of content
        content = content.text;
      }
      if (content?.length > 500) {
        content = content.slice(0, 500) + "â€¦";
      }

      // Whitelist what properties to include
      // Some sites include extra properties like comments that I don't need
      const { name, type, published, updated } = entry;

      references[url] = {
        url,
        name,
        type,
        published,
        updated,
        author,
        content,
      };
    }

    properties = {
      date: published,
      updated: properties.updated,
      // TODO: Check how visibility is set
      eleventyExcludeFromCollections: deleted
        ? "deleted"
        : properties["post-status"] === "draft"
        ? "draft"
        : properties.visibility === "public"
        ? undefined
        : properties.visibility,
      title: properties.name,
      slug: properties["mp-slug"],
      "bookmark-of": properties["bookmark-of"],
      "like-of": properties["like-of"],
      "repost-of": properties["repost-of"],
      "in-reply-to": properties["in-reply-to"],
      tags: properties.category,
      image,
      image_alt,
      references,

      // Untested properties
      image_other: Object.keys(image_other).length ? image_other : undefined,
      indiekit_excerpt: properties.summary,
      indiekit_start: properties.start,
      indiekit_end: properties.end,
      indiekit_rsvp: properties.rsvp,
      indiekit_location: properties.location,
      indiekit_checkin: properties.checkin,
      indiekit_audio: properties.audio,
      indiekit_video: properties.video,

      "indiekit_post-type": properties["post-type"],
      syndication: properties.syndication,
      "mp-syndicate-to": properties["mp-syndicate-to"],
    };

    return `---
${stringify(properties)}
---

${postContent}
`;
  } catch (error) {
    console.error("Problem generating post file content", error, properties);
    throw new Error(
      "Unable to create post file contents. See logs for details"
    );
  }
};
