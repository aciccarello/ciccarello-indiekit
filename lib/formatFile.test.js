import { jest } from "@jest/globals";
import formatFile from "./formatFile.js";

beforeAll(() => {
  jest.spyOn(console, "log").mockReturnValue();
});

afterAll(() => {
  jest.restoreAllMocks();
});

it("should create a photo post", async () => {
  expect(
    await formatFile((await import("./fixtures/photo-post.json")).default)
  ).toMatchSnapshot();
});

it("should ignore properties from Aaron's post", async () => {
  expect(
    await formatFile(
      (
        await import("./fixtures/aaronparecki_com_lawyer.json")
      ).default
    )
  ).toMatchSnapshot();
});

it("should process find the author in james' post", async () => {
  expect(
    await formatFile(
      (
        await import("./fixtures/jamesg_blog_folder_names.json")
      ).default
    )
  ).toMatchSnapshot();
});

it("should process David's checkin", async () => {
  expect(
    await formatFile(
      (
        await import("./fixtures/david_shanske_com_checkin.json")
      ).default
    )
  ).toMatchSnapshot();
});

it("should modify the content from marty's post", async () => {
  expect(
    await formatFile(
      (
        await import("./fixtures/martymcgui_re_go-time.json")
      ).default
    )
  ).toMatchSnapshot();
});

it("should fill in references from mastodon", async () => {
  expect(
    await formatFile(
      (
        await import("./fixtures/nick_nisi_mastodon.json")
      ).default
    )
  ).toMatchSnapshot();
});

it("should create a listen-of post from an indiekit jam", async () => {
  expect(
    await formatFile((await import("./fixtures/indiekit_jam.json")).default)
  ).toMatchSnapshot();
});

it("should create a listen-of post from sparkles", async () => {
  expect(
    await formatFile((await import("./fixtures/sparkles_listen.json")).default)
  ).toMatchSnapshot();
});

it("should create a read-of post from sparkles", async () => {
  expect(
    await formatFile((await import("./fixtures/sparkles_read.json")).default)
  ).toMatchSnapshot();
});
