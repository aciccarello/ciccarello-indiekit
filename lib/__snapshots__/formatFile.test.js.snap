// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a photo post 1`] = `
"---
date: 2023-11-09T10:10:00.000Z
eleventyExcludeFromCollections: true
eleventyExcludeFromCollectionsReason: draft
slug: noyzk
image: /assets/img/8pzun.jpg
image_alt: pumpkin bread on cutting board
indiekit_post-type: photo
---

Draft post

"
`;

exports[`should fill in references from mastodon 1`] = `
"---
date: 2023-03-09T03:23:02Z
slug: sometimes-slack-is-fun
like-of: https://fediverse.nicknisi.com/@nicknisi/109991177881690708
references:
  - referenceIdUrl: https://fediverse.nicknisi.com/@nicknisi/109991177881690708
    url: https://fediverse.nicknisi.com/@nicknisi/109991177881690708
    type: entry
    author:
      type: card
      name: "Nick Nisi :nicknisi:"
      nickname: "@nicknisi@nicknisi.com"
      url: https://fediverse.nicknisi.com/@nicknisi
    content: |-
      Attached: 1 image

      Sometimes Slack is fun.
indiekit_post-type: like
---


"
`;

exports[`should ignore properties from Aaron's post 1`] = `
"---
date: 2023-04-28T06:39:26.637Z
eleventyExcludeFromCollections: true
eleventyExcludeFromCollectionsReason: draft
slug: aaron-lawyer
like-of: https://aaronparecki.com/2023/04/24/8/lawyer
references:
  - referenceIdUrl: https://aaronparecki.com/2023/04/24/8/lawyer
    url: https://aaronparecki.com/2023/04/24/8/lawyer
    type: entry
    published: 2023-04-24T14:12:20-07:00
    author:
      type: card
      url: https://aaronparecki.com/
      photo:
        - alt: Aaron Parecki
          url: https://aaronparecki.com/images/profile.jpg
      name: Aaron Parecki
    content: In retrospect, I probably didn't need to include "but I am not a
      lawyer" in an email to our lawyers
indiekit_post-type: like
---


"
`;

exports[`should modify the content from marty's post 1`] = `
"---
date: 2023-05-01T00:03:00.564Z
title: Go time time
slug: go-again-time
bookmark-of: https://martymcgui.re/2023/03/19/go-time/?utm_medium=rss&utm_source=rss
references:
  - referenceIdUrl: https://martymcgui.re/2023/03/19/go-time/?utm_medium=rss&utm_source=rss
    url: https://martymcgui.re/2023/03/19/go-time/
    name: Go Time
    type: entry
    published: 2023-03-19T13:14:22-0400
    author:
      type: card
      name: Marty McGuire
      url: https://martymcgui.re/
      photo:
        - alt: Marty McGuire
          url: https://martymcgui.re/images/logo.jpg
    content: >-
      This is one of those posts about a tech issue comes up from time to time
      for me that I find difficult to search for. Sorry (not sorry) if it is not
      helpful to anyone else!

        Too much context about my site building process

        

      My site is built with Hugo, a static site generator known for being quite
      fast, and with a powerful (and challenging) templating system.

      I use my site for lots of stuff, which sometimes means showing content
      from other sites in my posts. For example:

        
          Responses that…
indiekit_post-type: bookmark
---


"
`;

exports[`should process David's checkin 1`] = `
"---
date: 2023-04-28T06:46:25.459Z
eleventyExcludeFromCollections: true
eleventyExcludeFromCollectionsReason: draft
title: David's Checkin
slug: davids-checkin
bookmark-of: https://david.shanske.com/2022/11/12/5771/
references:
  - referenceIdUrl: https://david.shanske.com/2022/11/12/5771/
    url: https://david.shanske.com/2022/11/12/5771/
    post-type: checkin
    type: entry
    published: 2022-11-12T21:18:14-05:00
    author:
      type: card
      name: David Shanske
      url: https://david.shanske.com/
      photo: https://david.shanske.com/avatar/dshanske?s=42
indiekit_post-type: bookmark
---


"
`;

exports[`should process find the author in james' post 1`] = `
"---
date: 2023-04-29T06:00:13.378Z
eleventyExcludeFromCollections: true
eleventyExcludeFromCollectionsReason: draft
slug: abcd
like-of: https://jamesg.blog/2023/04/18/source-code-folder-names/
references:
  - referenceIdUrl: https://jamesg.blog/2023/04/18/source-code-folder-names/
    url: https://jamesg.blog/2023/04/18/source-code-folder-names/
    name: My source code root folder name
    type: entry
    published: 2023-04-18T00:00:00
    author:
      type: card
      name: James' Coffee Blog ☕
      url: https://jamesg.blog
    content: >-
      I like seeing what people call the root folder in which they store their
      source code. This is the folder where all — or a lot of — your projects
      are stored. In my case, my programming projects go in a folder called src.
      (Although I have a strange habit of nesting personal projects that are
      related to each other. I believe my source code files are in need of a
      spring clean.)

      That long parenthetical notwithstanding, I find the name src cool. It’s a
      short way of saying source code; apt, simple, eas…
indiekit_post-type: like
---


"
`;
